docs/source/jit_language_reference_v2.rst

```diff
@@ -0,0 +1,320 @@
+.. contents::
+    :local:
+    :depth: 2
+
+
+.. testsetup::
+
+    # These are hidden from the docs, but these are necessary for `doctest`
+    # since the `inspect` module doesn't play nicely with the execution
+    # environment for `doctest`
+    import torch
+
+    original_script = torch.jit.script
+    def script_wrapper(obj, *args, **kwargs):
+        obj.__module__ = 'FakeMod'
+        return original_script(obj, *args, **kwargs)
+
+    torch.jit.script = script_wrapper
+
+    original_trace = torch.jit.trace
+    def trace_wrapper(obj, *args, **kwargs):
+        obj.__module__ = 'FakeMod'
+        return original_trace(obj, *args, **kwargs)
+
+    torch.jit.trace = trace_wrapper
+
+.. _language-reference:
+
+TorchScript Language Reference
+==============================
+
+.. _expressions:
+
+
+Expressions
+~~~~~~~~~~~
+
+The following section describes the grammar of expressions that are supported in TorchScript.
+It is modeled after `the expressions chapter of the Python language reference <https://docs.python.org/3/reference/expressions.html>`_.
+
+Arithmetic Conversions
+^^^^^^^^^^^^^^^^^^^^^^
+There are a number of implicit type conversions that are performed in TorchScript:
+
+
+* a ``Tensor`` with a ``float`` or ``int`` datatype can be implicitly converted to an instance of ``FloatType`` or ``IntType`` provided that it has a size of 0, and does not have ``require_grad`` set to ``True`` and will not require narrowing.
+* instances of ``StringType`` can be implicitly converted to ``DeviceType``
+* the above implicit conversion rules can be applied to instances of ``TupleType`` to produce instances of ``ListType`` with the appropriate contained type
+
+
+Explicit conversions can be invoked using the ``float``, ``int``, ``bool``, ``str`` built-in functions
+that accept primitive data types as arguments and can accept user-defined types if they implement
+``__bool__``, ``__str__``, etc.
+
+
+Atoms
+^^^^^
+Atoms are the most basic elements of expressions.
+
+::
+
+    atom      ::=  identifier | literal | enclosure
+    enclosure ::=  parenth_form | list_display | dict_display
+
+Identifiers
+"""""""""""
+The rules that dictate what is a legal identifer in TorchScript are the same as
+their `Python counterparts <https://docs.python.org/3/reference/lexical_analysis.html#identifiers>`_.
+
+Literals
+""""""""
+
+::
+
+    literal ::=  stringliteral | integer | floatnumber
+    `stringliteral <https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals>`
```

Comment:
It seems this still doesn't render as links, maybe move it outside of code block as we discussed last time?